### PR TITLE
return proper error if posting project with the same name

### LINKED ISSUE #195: Raise better error when creating duplicate projects
https://github.com/osmlab/to-fix-backend/blob/master/routes/project.js#L72

We should catch a UniqueContraintValidation error or so, just like https://github.com/osmlab/to-fix-backend/blob/master/routes/item.js#L271 and inform the user that project creation failed because it already exists.

cc @emilymdubois

### PR DESCRIPTION
**Pull request by batpad**
Fixes #195 

cc @emilymdubois

### CODE REVIEW

**Comment by batpad** [comment]
@emilymdubois sorry, overloaded this PR a bit with fix for #194 - I think we really need some way to do this -- the current way I've implemented it is probably suboptimal, but hopefully it's fine for now and we can revisit later.

File: routes/project.js [hunk 1]
```diff
@@ -31,7 +33,16 @@ module.exports = {
  * ]
  */
 function getProjects(req, res, next) {
-  Project.findAll()
+  let search = {};
+
+  //FIXME: this is probably not the best way to implement this
+  // since a search for a name should always return a maximum of 1 item
```
Comment (emilymdubois) [inline routes/project.js:39]: There may be function that's a happy medium (does `.find` exist?) but I agree that this works for now and we can always revisit later.

File: routes/project.js [hunk 2]
```diff
@@ -17,6 +17,8 @@ module.exports = {
 /**
  * Get a list of projects.
  * @name get-projects
+ * @param {Object} query - The request URL query parameters
+ * @param {string} query.name - Name of project to filter by (optional)
```
Comment (emilymdubois) [inline routes/project.js]: You can actually connote optional parameters with square brackets; for example:

https://github.com/osmlab/to-fix-backend/blob/10e3e31ae2a72f8928a559b9bc6495d5a804e4a2/routes/tag.js#L32-L33

documentation.js will automatically add "optional" to the documentation.

File: routes/project.js [hunk 3]
```diff
@@ -31,7 +33,16 @@ module.exports = {
...
  */
 function getProjects(req, res, next) {
-  Project.findAll()
+  let search = {};
+
+  //FIXME: this is probably not the best way to implement this
+  // since a search for a name should always return a maximum of 1 item
+  if (req.query.name) {
+    search.where = {
+      name: req.query.name
```
Comment (emilymdubois) [inline routes/project.js:42]: What if the project name has a space in it? Does this get interpreted in `req.query.name` as something like `My%20Project`?  If so, here's an example of how I implemented handling around URL strings:

https://github.com/osmlab/to-fix-backend/blob/10e3e31ae2a72f8928a559b9bc6495d5a804e4a2/routes/tag.js#L51-L54

File: routes/project.js [hunk 4]
```diff
@@ -69,7 +80,13 @@ function createProject(req, res, next) {
     .then(function(data) {
       res.json(data);
     })
-    .catch(next);
+    .catch(err => {
+      if (err instanceof Sequelize.UniqueConstraintError) {
+        return next(new ErrorHTTP('Project with name already exists', 400));
```
Comment (emilymdubois) [inline routes/project.js:85]: This is nice 😌

**Review by emilymdubois — COMMENTED** [review]
There's only one comment in here about spaces in project names that may need addressing; everything else is stylistic!

File: routes/project.js (hunk 3 above)
Comment (batpad) [reply routes/project.js:42]: So I thought of this, but the test that I have does search for a project with a space in its name, and it works fine.

From my experience with frameworks, they should always handle query string decoding for you transparently - so when making the query, of course, you need to `encodeURIComponent` your query strings, but when reading it out in express, express should just handle this for you internally, and as per the passing tests we have, that works :-) - fine to leave as is?

Comment (emilymdubois) [reply routes/project.js:42]: 👍  Sounds good!

**Pull request merged**